Calculate task metric aggregates on-the-fly to reduce memory usage

sayedbilalbari opened the pull request:
Signed off by: Sayed Bilal Bari ( sayedbilalbari ) 

Fixes #1545 

This PR updates the AccumInfo logic to not store the updates associated with each task and maintain rolling Statistics at the stage level. This improves the overall memory usage by qualification run.

## Why this change?

- This change aims to **reduce overall heap usage** of the tools qualification run.
- The qualification tool **parses Spark event logs**, collates data in memory, and outputs aggregate metrics (SQL metrics, Job level metrics, etc.) to generate a recommendation report.
- **Major bottleneck** identified:
  - The tool stores **all** update values coming in for each Task towards an Accumulable.
  - Because **Tasks** are numerous, the `HashMap` storing these update values can become **very large**. 
    - For example, out of 14G of usage for an event log, the update map uses ~10G.
  - These update values are primarily used to calculate **min, median, max,** and **total** contribution of a stage towards an Accumulable (e.g., total shuffle).
- Since the **final usage** is just min, median, max, and total, this PR changes the previous behavior:
  - **On-the-fly** calculation of these metrics without storing all update values in memory.

## Change

1. **`AccumInfo`**:
   - Updated to **remove** the `taskUpdateMap`.
   - The `stageUpdateMap` now:
     - Stores a map of values associated with each task.
     - Maintains a `StatisticMetric` object that is **updated with the latest task metrics** on a rolling basis.
2. **`AppSQLPlanAnalyzer`**:
   - Previously, there was significant aggregation at the **stage** level.
   - Updated to use a **new method** for calculating metrics at the stage level.
   - **Removed** other methods to favor the `StageLevel` metrics object.
3. **`Profiler`**:
   - Timeline generation is **no longer used** - will be tackled in next PR
   - Updated code for `TimelineGeneration` to remove taskUpdate usage

## Testing

- The **output reports** generated by this change were compared against the original output reports:
  - **No change** in the recommendations.
  - The **only** difference in raw metrics is in **median** values, since a rolling average is now used instead of a strict median.
    - Observed no major outliers or disparities.
- Local test cases failing only due to the slight **change in median** values.
  - Very small difference observed.

## Improvement

Metrics are for a **34G Databricks ZSTD log**:

- **Total runtime** improvement of **37%**  
- **Peak heap usage** improvement of **68%**   from **15G** to **5G**.
- **Min heap requirement** improvement of **85%** from **14G** to **3G**

![Screenshot 2025-02-10 at 5 32 33 PM](https://github.com/user-attachments/assets/d67a6d53-f09e-462b-942d-febf412ed5e1)  
![Screenshot 2025-02-13 at 10 43 55 AM](https://github.com/user-attachments/assets/fe927301-1d1e-459e-b454-144a75ab7af8)

**BenchmarkSuite output before change**: 

```
Benchmark :                               Best Time(ms)   Avg Time(ms)   Stdev(ms)      Avg GC Time(ms)       Avg GC Count     Stdev GC Count    Max GC Time(ms)       Max GC Count   Relative
--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
Enable_Per_SQL_Arg_Qualification                 722587         724549        2774              20098.0              364.0                 63              20601                396      1.00X
```
**BenchmarkSuite output after change**: 

```
Benchmark` :                               Best Time(ms)   Avg Time(ms)   Stdev(ms)      Avg GC Time(ms)       Avg GC Count     Stdev GC Count    Max GC Time(ms)       Max GC Count   Relative
--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
Enable_Per_SQL_Arg_Qualification                 449096         450413        1862               2951.0              198.0                  1               2955                199      1.00X
```

**Metrics by running with dev checkpoint enabled**
Ran with JVM params ->
-Xms=16G
-Xmx=16G

Before -
```
Memory Marker: Post processing events, 
jvm.heap.max -> 17.17G
jvm.heap.total -> 17.17G
jvm.heap.free -> 4.2G
```
After -
```
Memory Marker: Post processing events, 
jvm.heap.max -> 17.17G
jvm.heap.total -> 17.17G
jvm.heap.free -> 14.77G
```

This pull request includes multiple changes aimed at improving the performance and accuracy of the Spark RAPIDS tool. The most significant changes involve the removal of unused methods, updates to the handling of accumulator statistics, and improvements to the profiling and analysis components.

### Codebase simplification and removal of unused methods:

* Removed the `getStageTaskIds` and `filterAccumTaskUpdatesForStage` methods from the `AppSQLPlanAnalyzer` class as they are no longer needed. [[1]](diffhunk://#diff-d56ed990ec181a2ef118996b608fdff3b53818cf6138baba59b51e4eb407d79eL111-L136) [[2]](diffhunk://#diff-d56ed990ec181a2ef118996b608fdff3b53818cf6138baba59b51e4eb407d79eL409) [[3]](diffhunk://#diff-d56ed990ec181a2ef118996b608fdff3b53818cf6138baba59b51e4eb407d79eL429-R402) [[4]](diffhunk://#diff-d56ed990ec181a2ef118996b608fdff3b53818cf6138baba59b51e4eb407d79eL480-R453)

### Enhancements to accumulator statistics:

* Updated the `AccumInfo` class to use `stageValuesMap` for storing stage-level metrics and added methods to calculate and retrieve statistics. This change improves the accuracy of the accumulator statistics by leveraging stage-level metrics. [[1]](diffhunk://#diff-2cdf5cec29c5cfc15962382b2134c8e88b6623afdfd7cc6a81ec3dfc5663b4a1L2-R2) [[2]](diffhunk://#diff-2cdf5cec29c5cfc15962382b2134c8e88b6623afdfd7cc6a81ec3dfc5663b4a1L19-R19) [[3]](diffhunk://#diff-2cdf5cec29c5cfc15962382b2134c8e88b6623afdfd7cc6a81ec3dfc5663b4a1L35-R161)
* Modified the `StatisticsMetrics` case class to use mutable fields for `max` and `total` to support incremental updates.

### Profiling and analysis improvements:

* Updated the `GenerateTimeline` object to use `stageValuesMap` for calculating semaphore wait times and other metrics, improving the accuracy of the profiling data. [[1]](diffhunk://#diff-1a2aa4edeab8f473a1a10f8fb6d0e676a0537197bfd1fd26fd09a58c8538a233L279-L293) [[2]](diffhunk://#diff-1a2aa4edeab8f473a1a10f8fb6d0e676a0537197bfd1fd26fd09a58c8538a233L303-R299)
* Adjusted the `AppSparkMetricsAnalyzer` class to leverage stage-level metrics for peak memory and shuffle write time calculations, enhancing the profiling of Photon applications.

### Miscellaneous updates:

* Updated the copyright year in several files to 2025. [[1]](diffhunk://#diff-13353883e8b8763b970c1ad85d6c670336d17e492997f1cfb812902b2c436a6cL2-R2) [[2]](diffhunk://#diff-2cdf5cec29c5cfc15962382b2134c8e88b6623afdfd7cc6a81ec3dfc5663b4a1L2-R2)
* Modified the profiling expectations CSV file to reflect the changes in the accumulator statistics.

These changes collectively improve the performance, accuracy, and maintainability of the Spark RAPIDS tool.

amahussein commented:
> Profiler - timeline generation is no longer used. Removed code for TimelineGeneration
was it possible to do some workarounds to keep that functionality for now? if we can avoid removing the feature in this PR it will be better. Otherwise, we will have to fix the CI/CDs and the QA tests
> Metrics mentioned are for a 34G databricks zstd log
> Total runtime comes down 27% ( from 9 min 20 sec to 6 min 50 sec )
> Memory utilisation comes down 68% ( from 15G to 5G )

Is the runtime evaluation done through the benchmarkSuite?

What is meant here by "Memory utilisation"? Does it mean minimum required heap to run the eventlog? or the total allocated bytes as measured by the Java's profiler?

amahussein commented on core/src/main/scala/org/apache/spark/sql/rapids/tool/store/AccumInfo.scala:
```diff
@@ -1,5 +1,5 @@
 /*
- * Copyright (c) 2024, NVIDIA CORPORATION.
+ * Copyright (c) 2025, NVIDIA CORPORATION.
```
`2024-2025`

sayedbilalbari replied:
Updated

amahussein commented on core/src/main/scala/com/nvidia/spark/rapids/tool/profiling/GenerateTimeline.scala:
```diff
@@ -132,474 +132,474 @@ object GenerateTimeline {
+//    } finally {
+//      fileWriter.close()
+//    }
+//  }
```
If we have to remove that generateTimeLine feature, then perhaps we break the PR into 2. One to remove the generateTimeLine.
This will make the PR manageable, easier to read and also separate the followups that we have to deal with from removing this unrelated feature. (this requires creating a new issue)

sayedbilalbari replied:
Currently I have updated the GenerateTimeline code to not use the TaskUpdateMap ( lose sub timing precise information )- 
- `readTimeMs` - it maintained extra information regarding read accumulable
- `opMetrics` - it maintained extra information regarding op accumulable
- `writeMetric` - it maintained extra information regarding write accumulable

Have removed the extra time addition. We can tackle this in a separate PR where we either get rid of this or figure out a precise calculation mechanism for it

amahussein commented on core/src/main/scala/org/apache/spark/sql/rapids/tool/store/AccumInfo.scala:
```diff
@@ -32,11 +32,10 @@ import org.apache.spark.sql.rapids.tool.util.EventUtils.parseAccumFieldToLong
-    new mutable.HashMap[Long, Long]()
-  val stageValuesMap: mutable.HashMap[Int, Long] =
-    new mutable.HashMap[Int, Long]()
+  // TODO: use sorted maps for stageIDs and taskIds
```
The todo needs to be updated.

sayedbilalbari replied:
Updated

sayedbilalbari commented:
@amahussein Thanks for the review !
- Have updated the code to keep GenerateTimeline as of now
- Added benchmarkSuite runtime and updated the exact improvement metric
- Have added precise memoryUtilisation metrics as well ( peak heap usage and min heap usage )

amahussein commented on core/src/main/scala/com/nvidia/spark/rapids/tool/analysis/StatisticsMetrics.scala:
```diff
@@ -1,5 +1,5 @@
 /*
- * Copyright (c) 2024, NVIDIA CORPORATION.
+ * Copyright (c) 2025, NVIDIA CORPORATION.
```
`2024-2025` :)

The copyrights year is working as follows:

- A new file gets the year when it is created for the first time.
- When a file is modified, the copyrights year is updated by setting the range of the year the file is created ending up by the latest modification year.

The pre-commit hooks defined in this repo should help automating that step.

sayedbilalbari replied:
Updated, thank you ! Will ensure this next time

amahussein reviewed (commented):
Thanks @sayedbilalbari 
This is going to be a great milestone in the optimizations.

There is a few things we need to address:

- When evaluating the memory savings, I like to see using the `dev` mode to show the heap size after reading the eventlog. Then we compare that with the dev branch. This method is way more reliable and consistent. It helps us to set a reference for the subsequent changes in the tools metrics. In addition, not all devs/reviewers would be using the same Java profiler tool. While the JProfiler is handy, we want to be able to converge to standard means to evaluate performance.
  - It is a bonus if there is a script that builds in `dev` mode to gather these numbers, but I understand if we want to keep that in a different PR.
- We want to review how incomplete stages are handled. IIRC, taskEnd events are triggered before the stageCompletion. Also, we create stageModel only during the handling of stageSubmitted. We can do that in an offline discussion.
- since we are doing average instead of median. Probably we may consider renaming the fields to average instead.  we can also discuss that offline.
-

sayedbilalbari commented on core/src/main/scala/com/nvidia/spark/rapids/tool/profiling/GenerateTimeline.scala:
```diff
@@ -60,7 +60,7 @@ object GenerateTimeline {
   private val MS_PER_PIXEL = 5.0
 
-  // Generated using https://mokole.com/palette.html
+//   Generated using https://mokole.com/palette.html
```
Corrected

sayedbilalbari commented on core/src/test/resources/ProfilingExpectations/rapids_join_eventlog_sqlmetrics_expectation.csv:
```diff
@@ -20,65 +20,65 @@ appIndex,sqlID,nodeID,nodeName,accumulatorId,name,min,median,max,total,metricTyp
 1,0,3,"GpuColumnarExchange",56,"shuffle bytes written",77,77,77,15400,"size","2,3"
 1,0,3,"GpuColumnarExchange",57,"shuffle records written",1,1,1,200,"sum","2,3"
-1,0,3,"GpuColumnarExchange",58,"shuffle write time",139875,230038,9747416,93193331,"nsTiming","2,3"
+1,0,3,"GpuColumnarExchange",58,"shuffle write time",139875,465911,9747416,93193331,"nsTiming","2,3"
```
Yes correct. There is a considerable change in the median value in some cases ( expected ). 
The closest metric to that is the rolling average which this PR uses.
We can update the column name, but my doubt is that in that case do we also need to update the raw metric files header and how will that impact the downstream qualX consuming those metrics.
We can discuss this offline

sayedbilalbari replied:
We can even take this up in a separate PR

sayedbilalbari commented:
@amahussein Thanks for the review again !
- Have added the metrics from running the dev checkpoint. Added it to the description. Regarding the script, we can add that in a separate PR
- For handling missing StageCompleted events, for each TaskEnd, the stageUpdate map is updated with the cummulative update values.
- We create the StageModel when we get a StageSubmitted and update that on StageCompleted.
- If we miss out on the StageCompleted event, the metric that we might miss out on would be the completionTime, rest should be updated with the StageSubmitted event

amahussein commented on core/src/main/scala/org/apache/spark/sql/rapids/tool/store/AccumInfo.scala:
```diff
@@ -32,81 +32,133 @@ import org.apache.spark.sql.rapids.tool.util.EventUtils.parseAccumFieldToLong
+   * - Total accumulator value for the stage (Long)
+   * - Statistical metrics for task-level updates (StatisticsMetrics)
+   */
+  val stageValuesMap: mutable.HashMap[Int, (Long, StatisticsMetrics)] =
```
Perhaps better to replace that tuple with a proper class. It is not that easily readable/maintainable to use that tupe especially it is a milestone for the metrics accumulation.
Instead of `_1` and `_2` we would have field name that makes it easier to understand the code.

amahussein replied:
Another question. if we have statisticsMetrics.total field..then why don't we use that to represent the total? And use the `_1` as a counter?

sayedbilalbari replied:
Have updated the structure to incorporate the count and total as part of the StatisticMetric.

amahussein commented on core/src/main/scala/org/apache/spark/sql/rapids/tool/store/AccumInfo.scala:
```diff
@@ -32,81 +32,133 @@ import org.apache.spark.sql.rapids.tool.util.EventUtils.parseAccumFieldToLong
       case _ => update.getOrElse(0L)
     }
-    stageValuesMap.put(stageId, Math.max(existingValue, incomingValue))
+    val newValue = Math.max(existingEntry._1, incomingValue)
```
add more code comments here to explain why we take the max and any other necessary details.

amahussein commented on core/src/main/scala/org/apache/spark/sql/rapids/tool/store/AccumInfo.scala:
```diff
@@ -32,81 +32,133 @@ import org.apache.spark.sql.rapids.tool.util.EventUtils.parseAccumFieldToLong
-   * @param accumulableInfo the accumulableInfo from the TaskEnd event
+   * Processes task-level accumulator updates and updates stage-level statistics.
+   * Called during TaskEnd event processing.
+   *
```
same comment here about "attempts". Should we add arguments for stageAttempt and task attempt? If we do, then lets put a TODO. If not, then explain why.
Then any one reading that code won't have to dig into whether the attempt id is required or not.

sayedbilalbari replied:
Updated with explanation

amahussein commented on core/src/main/scala/org/apache/spark/sql/rapids/tool/store/AccumInfo.scala:
```diff
@@ -32,81 +32,133 @@ import org.apache.spark.sql.rapids.tool.util.EventUtils.parseAccumFieldToLong
    */
   def addAccumToTask(stageId: Int, taskId: Long, accumulableInfo: AccumulableInfo): Unit = {
+    // For each incoming Task, we update the statistics with min, rolling average
+    // max and count of tasks
```
Is this an incomplete sentence? 
For example `first step: Calculate count of tasks and max updateValue`

sayedbilalbari replied:
Made it more verbose

amahussein commented on core/src/main/scala/org/apache/spark/sql/rapids/tool/store/AccumInfo.scala:
```diff
@@ -32,81 +32,133 @@ import org.apache.spark.sql.rapids.tool.util.EventUtils.parseAccumFieldToLong
-    // update the stage value map if necessary
-    if (updateStageFlag) {
-      addAccumToStage(stageId, accumulableInfo, parsedUpdateValue)
+    parsedUpdateValue.foreach{ value =>
```
missing white space before curley braces..I thought I added scala checkstyle to detect that.

sayedbilalbari replied:
Updated. This did not pop up in verify check

amahussein commented on core/src/main/scala/org/apache/spark/sql/rapids/tool/store/AccumInfo.scala:
```diff
@@ -32,81 +32,133 @@ import org.apache.spark.sql.rapids.tool.util.EventUtils.parseAccumFieldToLong
+        Math.max(stats.max, value),
+        stats.total + 1
+      )
+      stageValuesMap.put(stageId, (total + value, newStats))
```
put more code explanation of what is going on there. For example how the rolling average is calculated.
Also will be nice to use the argument so that it is clear which value is being assigned to which field of the statics

sayedbilalbari replied:
Added explanation.

amahussein commented on core/src/main/scala/org/apache/spark/sql/rapids/tool/store/AccumInfo.scala:
```diff
@@ -32,81 +32,133 @@ import org.apache.spark.sql.rapids.tool.util.EventUtils.parseAccumFieldToLong
+        Math.min(stats.min, value),
+        (stats.med * stats.total + value) / ( stats.total + 1),
+        Math.max(stats.max, value),
+        stats.total + 1
```
put a comment that his represents the count to be used for the rolling average. It is still can be confusing because the field name is `total`.

sayedbilalbari replied:
Changed implementation to make it more understandable

amahussein reviewed (changes requested):
Overall. looks great!
Some comments to make it even better :)

sayedbilalbari commented on core/src/main/scala/com/nvidia/spark/rapids/tool/analysis/AppSparkMetricsAnalyzer.scala:
```diff
@@ -399,15 +399,15 @@ class AppSparkMetricsAnalyzer(app: AppBase) extends AppAnalysisBase(app) {
+        // We leverage the stage level metrics and get the max task update from it
+        val peakMemoryValues = photonPeakMemoryAccumInfos.flatMap { accumInfo =>
+          accumInfo.stageValuesMap.get(sm.stageInfo.stageId)
+        }.map(_._2.max)
```
Updated

sayedbilalbari commented on core/src/main/scala/com/nvidia/spark/rapids/tool/analysis/AppSparkMetricsAnalyzer.scala:
```diff
@@ -399,15 +399,15 @@ class AppSparkMetricsAnalyzer(app: AppBase) extends AppAnalysisBase(app) {
+          accumInfo.stageValuesMap.get(sm.stageInfo.stageId)
+        }.map(_._2.max)
         // For sum of shuffle write time, we need to look at the accumulators at the stage level.
+        // We get the values associated with all tasks for it
```
Updated

sayedbilalbari commented on core/src/main/scala/com/nvidia/spark/rapids/tool/analysis/AppSparkMetricsAnalyzer.scala:
```diff
@@ -399,15 +399,15 @@ class AppSparkMetricsAnalyzer(app: AppBase) extends AppAnalysisBase(app) {
         val shuffleWriteValues = photonShuffleWriteTimeAccumInfos.flatMap { accumInfo =>
           accumInfo.stageValuesMap.get(sm.stageInfo.stageId)
-        }
+        }.map(_._1)
```
Updated

sayedbilalbari commented on core/src/main/scala/com/nvidia/spark/rapids/tool/profiling/GenerateTimeline.scala:
```diff
@@ -276,21 +276,15 @@ object GenerateTimeline {
     val semMetricsNs = semWaitIds.toList
       .flatMap(app.accumManager.accumInfoMap.get)
-      .flatMap(_.taskUpdatesMap.values).sum
+      .flatMap(_.stageValuesMap.values).map(_._1).sum
```
Updated

sayedbilalbari commented on core/src/main/scala/com/nvidia/spark/rapids/tool/profiling/GenerateTimeline.scala:
```diff
@@ -276,21 +276,15 @@ object GenerateTimeline {
         case (_, accumInfo: AccumInfo)
             if accumInfo.infoRef.name == AccumNameRef.NAMES_TABLE.get("gpuSemaphoreWait") =>
-            Some(accumInfo.taskUpdatesMap.values.sum)
+            Some(accumInfo.stageValuesMap.values.map(_._1).sum)
```
Updated

sayedbilalbari commented on core/src/main/scala/org/apache/spark/sql/rapids/tool/store/AccumInfo.scala:
```diff
@@ -32,81 +32,133 @@ import org.apache.spark.sql.rapids.tool.util.EventUtils.parseAccumFieldToLong
+   *
+   * @param stageId The ID of the stage
+   * @param accumulableInfo Accumulator information from the event
+   * @param update Optional task update value for TaskEnd events
```
Updated

sayedbilalbari commented on core/src/main/scala/org/apache/spark/sql/rapids/tool/store/AccumInfo.scala:
```diff
@@ -32,81 +32,133 @@ import org.apache.spark.sql.rapids.tool.util.EventUtils.parseAccumFieldToLong
     // lower-than the previous value. In that case we should take the maximum.
-    val existingValue = stageValuesMap.getOrElse(stageId, 0L)
+    val existingEntry = stageValuesMap.getOrElse(stageId,
+      (0L, StatisticsMetrics(0L, 0L, 0L, 0L)))
```
Updated

amahussein commented on core/src/main/scala/com/nvidia/spark/rapids/tool/analysis/StatisticsMetrics.scala:
```diff
@@ -19,12 +19,12 @@ package com.nvidia.spark.rapids.tool.analysis
 
 // Store (min, median, max, total) for a given metric
-case class StatisticsMetrics(min: Long, med: Long, max: Long, total: Long)
+case class StatisticsMetrics(min: Long, med: Long, max: Long, count: Int, total: Long)
```
- styling issue for future reference: it is better practice to add the new field at the end, except when the fields are arranged in a specific order. No need to change it here though.
- If number of tasks in a stage is `Long`, then this `count` field should be `Long` as well.
- In the above class, it does not save us memory by setting the "count" to an integer. Because the size of objects in Java are multiple of 8 (for memory alignments). So, what java does, is that is going to ceil the size of that class to 8X leaving the last 4 bytes as a padding.

Another point of discussion: there was another option to create a different class that has (count, statistics) fields. I think the tradeoff between the 2 options depends whether "count" field is going to be used in other contexts. We know it is used in the Accumulable updates, but how much it is needed when we create AccumProfilResults. We can end up creating a ton of that class for the aggregation without using that field at all. 
On the other hand, it makes sense that "Statistics" should have a field for count.

sayedbilalbari replied:
Yeah, my thought process, it makes sense to have a rolling count of seen task events here and keep it as part of metrics.
Any access to it can happen through the AccumModel.
AccumProfile result has no count usage. 
I see count being of importance in StageDiagnosticResult but there we are using the logic of getting all tasks per stage and then using them.
Keeping it as part of statistics made more sense to me.

amahussein commented on core/src/main/scala/com/nvidia/spark/rapids/tool/profiling/GenerateTimeline.scala:
```diff
@@ -276,21 +276,15 @@ object GenerateTimeline {
     val semMetricsNs = semWaitIds.toList
       .flatMap(app.accumManager.accumInfoMap.get)
-      .flatMap(_.taskUpdatesMap.values).sum
+      .map(_.getTotalAcrossStages).sum
```
just verifying that it correct that we are doing `total.sum`?

sayedbilalbari replied:
Yes this is correct. 
This metric is calculating total semaphore wait time, which is the sum of all stage values for all all accumulables that match a regex. Here for all accumulables, we add the total time across stages and them sum them up.

amahussein reviewed (approved):
LGTME! Thanks @sayedbilalbari !
Lets move fwd with this to make it easy to investigate the correctness of metrics
